### ISSUE TITLE
`__init_subclass__` raising or annotated as returning `Never` isn't accounted for with subclass definitions

### ISSUE THREAD

**Issue by Sachaa-Thanasius** [microsoft/pyright#6858]
**Describe the bug**
Not entirely sure if this is a bug report or an enhancement request, honestly:

Let's say a class inherits from a mixin with a custom `__init_subclass__` that always raises an exception, and the method is annotated thusly with `Never` for the return type. However, pyright doesn't flag any subclass definitions as unreachable or disallowed (unlike with subclasses inheriting from a class decorated with `typing.final`) even though the subclass definition fails at runtime. Should this method's `Never`/`NoReturn` annotation be considered, or perhaps rather, should pyright more strictly analyze `__init_subclass__` behavior and flag it for things like this?

**Code or Screenshots**
Code:
```python
from typing import Never


class FinalMixin:
    def __init_subclass__(cls) -> Never:
        msg = "Subclassing FinalMixin is disallowed."
        raise TypeError(msg)


class Derived(FinalMixin):  # Not flagged by type-checker, though I instinctively expected it to.
    pass
```

Traceback:
```shell
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\Tushaar\Projects\scratch\py312_scratch\scratch2.py", line 10, in <module>
    class Derived(FinalMixin):
  File "C:\Users\Tushaar\Projects\scratch\py312_scratch\scratch2.py", line 7, in __init_subclass__
    raise TypeError(msg)
TypeError: Subclassing FinalMixin is disallowed.
```

**VS Code extension or command-line**
VS Code: Pylance v2023.12.101

**Comment by erictraut** [comment]
Even if this behavior were generally useful, it wouldn't be appropriate for a type checker to do what you're suggesting. It is not an error to call a function with a return type of `NoReturn` or `Never`. It simply means that control will not return back to the caller.
